app/ide-desktop/lib/content/src/utils.ts

```diff
@@ -0,0 +1,30 @@
+/** @file Module containing utility functions used throughout this module,
+ * but don't fit anywhere else. */
+
+/** Used to create a "branded type", which is a type intersected with a `Brand<'Name'>`.
+ *
+ * This is useful in parameters that require values from a certain source,
+ * for example IDs for a specific object type.
+ *
+ * It is similar to a `newtype` in other languages,
+ * however because TypeScript is structurally typed, a branded type is assignable to its base type:
+ * `a: string = b as (string & Brand<'Name'>)` is valid. */
+export interface Brand<T extends string> {
+    $brand: T
+}
+
+interface NoBrand {
+    $brand?: never
+}
+
+export function brand<T extends Brand<string>>(s: NoBrand & Omit<T, '$brand'>): T {
+    // This cast is unsafe. It is possible to use this method to cast a value from a base type to a
+    // branded type, even if that value is not an instance of the branded type. For example, the
+    // string "foo" could be cast to the `UserPoolId` branded type, although this string is clearly
+    // not a valid `UserPoolId`. This is acceptable because the branded type is only used to prevent
+    // accidental misuse of values, and not to enforce correctness.  That is, it is up to the
+    // programmer to declare the correct type of a value. After that point, it is up to the branded
+    // type to keep that guarantee by preventing accidental misuse of the value.
+    // eslint-disable-next-line no-restricted-syntax
+    return s as unknown as T
```

Comment:
1. Double space `[...] correctness.  That is [...]`
2. If this is unsafe, it should be named `unsafeBrand` to indicate in the usage place that its unsafe.
3. I don't like that it is so unsafe. Sure, it's up to the developer to use it correctly, but then, after code refactoring this could cause massive issues. If during refactoring we change the value type, this can hide serious issues. Is it possible to define it in as type-safe way somehow?

Reply:
the value type is required to be correct by the parameter's type: `NoBrand & Omit<T, '$brand'>`
(the `NoBrand` is there to disallow `Brand<"Foo"> & T` from being converted to `Brand<"Bar"> & T`

Reply:
The docs state that `[...] for example, string "foo" could be cast to the `UserPoolId` branded type [...]`. Tbh, I do not know what `Omit` means in TS, but if the docs are correct, it means that I could use this function to basically cast anything to anything, right? Or I misunderstood the docs?

Reply:
it's unsafe only in that strings don't have a property called `$brand` - this property that doesn't exist is used purely for extra type-safety

`Omit` leaves out a property - in this case it omits the `$brand` property from the *target type*, to obtain the original type (e.g. when given `ProjectId`, it outputs `string`, so that you can pass in a regular `string` to this function and get back a `ProjectId`)

Reply:
the docs do seem to be a bit misleading though, i can see whether i can make it clearer

Reply:
Ok, Id be thankful if you could make the docs clearer then and I'd love to re-read them and then continue this discussion, as I still don't fully understand when we could have non-type safe code when using it.